core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -90,14 +90,17 @@ public void onExpiration() {
      */
     @Override
     public void onComplete() {
+        // We are utilizing lock so that onComplete doesn't do a dirty read for global variables -
+        // partitionsAcquired and partitionsAlreadyFetched, since these variables can get updated in a different tryComplete thread.
+        lock.lock();
         log.trace("Completing the delayed share fetch request for group {}, member {}, "
             + "topic partitions {}", shareFetchData.groupId(), shareFetchData.memberId(),
             partitionsAcquired.keySet());
 
         if (shareFetchData.future().isDone())
             return;
```

Comment:
Hmm. Isn't this a return without unlocking? Seems wrong. I wonder if you should rearrange the code here so there's a try-finally and it's not possible to exit the method without unlocking.

Reply:
my bad, the PR was still a WIP and I opened it for review. Sorry about that. You're right, I am making changes along those lines right now.